portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java

```diff
@@ -0,0 +1,77 @@
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/nickname")
+public class NicknameServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html");
+    PrintWriter out = response.getWriter();
+    out.println("<h1>Set Nickname</h1>");
+
+    UserService userService = UserServiceFactory.getUserService();
+    if (userService.isUserLoggedIn()) {
+      String nickname = getUserNickname(userService.getCurrentUser().getUserId());
+      out.println("<p>Set your nickname here:</p>");
+      out.println("<form method=\"POST\" action=\"/nickname\">");
+      out.println("<input name=\"nickname\" value=\"" + nickname + "\" />");
+      out.println("<br/>");
+      out.println("<button>Submit</button>");
+      out.println("</form>");
+    } else {
+      String loginUrl = userService.createLoginURL("/nickname");
+      out.println("<p>Login <a href=\"" + loginUrl + "\">here</a>.</p>");
+    }
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/login");
+      return;
+    }
+
+    String nickname = request.getParameter("nickname");
+    String id = userService.getCurrentUser().getUserId();
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity entity = new Entity("UserInfo", id);
+    entity.setProperty("id", id);
+    entity.setProperty("nickname", nickname);
+    datastore.put(entity);
+
+    response.sendRedirect("/index.html");
+  }
+
+  /**
+   * Returns the nickname of the user with id, or empty String if the user has not set a nickname.
+   */
+  private String getUserNickname(String id) {
```

Comment:
DRY principle. You may extract this method in a separate class, for example `UserService`, and call the method here and in LoginServlet.